### PR TITLE
refactor/mvvm_messages_panel

### PR DESCRIPTION
**Pull request by teamagam1**

### CODE REVIEW

**Comment by review-ninja** [comment]
<a href="https://app.review.ninja/teamagam/GimelGimel/pull/108" target="_blank">ReviewNinja

File: domain/src/main/java/com/teamagam/gimelgimel/domain/messages/entity/MessageUserLocation.java [hunk 1]
```diff
@@ -3,15 +3,18 @@
...
 /**
  * UserLocation-Type class for {@link Message}
  */
 public class MessageUserLocation extends Message {
 
     private LocationSampleEntity mLocationSampleEntity;
 
-    public MessageUserLocation(String senderId, LocationSampleEntity sample) {
-        super(senderId);
+    public MessageUserLocation(String messageId, String senderId, Date createdAt, boolean isRead,
```
Comment (teamagam3) [inline domain/src/main/java/com/teamagam/gimelgimel/domain/messages/entity/MessageUserLocation.java]: Why did you change to constructor if you don't use the parameters in it?

File: domain/src/main/java/com/teamagam/gimelgimel/domain/messages/SendTextMessageInteractor.java [hunk 2]
```diff
@@ -25,6 +25,6 @@ protected SendTextMessageInteractor(
 
     @Override
     protected MessageText createMessage(String senderId) {
-        return new MessageText(senderId, mText);
+        return new MessageText(null, senderId, null, false, false, mText);
```
Comment (teamagam3) [inline domain/src/main/java/com/teamagam/gimelgimel/domain/messages/SendTextMessageInteractor.java:28]: If there's possibilty of putting a lot of nulls when constructing Message object, consider overloading the constructor

File: domain/src/main/java/com/teamagam/gimelgimel/domain/messages/SendImageMessageInteractor.java [hunk 3]
```diff
@@ -46,7 +46,7 @@ protected MessageImage createMessage(String senderId) {
         PointGeometry lastLocation = mLocationRepository.getLocation();
         ImageMetadata imageMetadata = new ImageMetadata(mImageTime, mImagesRepository.getImagePath(),
                 lastLocation, IMAGE_SOURCE);
-        return new MessageImage(senderId, imageMetadata);
+        return new MessageImage(null, senderId, null, false, false, imageMetadata);
```
Comment (teamagam3) [inline domain/src/main/java/com/teamagam/gimelgimel/domain/messages/SendImageMessageInteractor.java:49]: Same.

File: domain/src/main/java/com/teamagam/gimelgimel/domain/messages/SendGeoMessageInteractor.java [hunk 4]
```diff
@@ -59,18 +59,19 @@
 
     @Override
     protected MessageGeo createMessage(String senderId) {
-        PointSymbol symbol = createSymbolFromType(mMessageType);
+        PointSymbol symbol = createSymbolFromType(mMessageType, null);
         GeoEntity geoEntity = createGeoEntity(senderId + ":" + IdCreatorUtil.getUniqueId(),
                 mMessageGeometry,
                 symbol);
-        return new MessageGeo(senderId, geoEntity, mMessageText, mMessageType);
+        return new MessageGeo(null, senderId, null, false, false, geoEntity,
```
Comment (teamagam3) [inline domain/src/main/java/com/teamagam/gimelgimel/domain/messages/SendGeoMessageInteractor.java:66]: Same.

File: domain/src/main/java/com/teamagam/gimelgimel/domain/messages/SelectMessageInteractor.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+            String messageId) {
+        super(threadExecutor);
+        mMessagesRepository = messagesRepository;
+        mMessageId = messageId;
+    }
+
+
+    @Override
+    protected Observable<Message> buildUseCaseObservable() {
+        return mMessagesRepository.getMessageById(mMessageId)
```
Comment (teamagam3) [inline domain/src/main/java/com/teamagam/gimelgimel/domain/messages/SelectMessageInteractor.java:34]: Could be simplified by:

```
return mMEssageRepository.getMessageById(mMessageId).
                                             .doOnNext(mMessageRepository::selectMessage)
                                             .doOnNext(mMessageRepository::markMessageRead);
```

There is no reason to filter the messages, because we have unique id per message, and we want to select and read it, regarding the message is selected or read before.

File: domain/build.gradle [hunk 6]
```diff
@@ -25,4 +25,5 @@ dependencies {
     // auto-factory contains both annotations and their processor, neither is needed at runtime
     compileOnly "com.google.auto.factory:auto-factory:1.0-beta3"
     apt         "com.google.auto.factory:auto-factory:1.0-beta3"
+    compile 'org.jetbrains:annotations:15.0'
```
Comment (teamagam3) [inline domain/build.gradle:28]: Should we add new dependency just because we needed "@Nullable" annotation?

File: data/src/main/java/com/teamagam/gimelgimel/data/message/repository/MessagesDataRepository.java [hunk 7]
```diff
@@ -51,14 +63,53 @@ public MessagesDataRepository() {
...
+
+    @Override
+    public void selectMessage(Message message){
+        mCache.selectMessage(message);
+        mSelectedSubject.onNext(message);
+    }
+
+    @Override
+    public void markMessageRead(Message message){
+        mCache.markMessageRead(message);
```
Comment (teamagam3) [inline data/src/main/java/com/teamagam/gimelgimel/data/message/repository/MessagesDataRepository.java:102]: You have to message in this function, why do you pass the object to the cache?
The method in the InMemoryMessageCache only does message.setRead(true);

File: data/src/main/java/com/teamagam/gimelgimel/data/message/repository/MessagesDataRepository.java [hunk 8]
```diff
@@ -51,14 +63,53 @@ public MessagesDataRepository() {
...
+
+    @Override
+    public Observable<Message> getMessageById(String messageId) {
+        return Observable.just(messageId)
+                .map(mCache::getMessageById);
+    }
+
+    @Override
+    public void selectMessage(Message message){
+        mCache.selectMessage(message);
```
Comment (teamagam3) [inline data/src/main/java/com/teamagam/gimelgimel/data/message/repository/MessagesDataRepository.java:96]: Same as "marMessageRead()"

File: data/src/main/java/com/teamagam/gimelgimel/data/message/repository/InMemory/InMemoryMessagesCache.java [hunk 9]
```diff
@@ -3,26 +3,58 @@
...
+        }
+        return null;
+    }
+
+    public Observable<Collection<Message>> getMessages() {
+        return Observable.just(mMessages);
     }
 
+    public void markMessageRead(Message message) {
+        message.setRead(true);
```
Comment (teamagam3) [inline data/src/main/java/com/teamagam/gimelgimel/data/message/repository/InMemory/InMemoryMessagesCache.java:55]: Is it the cache job to do so?
Cache should save the data as is.

File: data/src/main/java/com/teamagam/gimelgimel/data/message/adapters/MessageDataMapper.java [hunk 10]
```diff
@@ -197,53 +158,47 @@ private MessageData transformToData(Message message) {
...
+        @Override
+        public void visit(MessageUserLocation message) {
+            LocationSampleData locationSampleData =
+                    mLocationSampleAdapter.transformToData(message.getLocationSample());
+            mMessageData = new MessageUserLocationData(locationSampleData);
         }
 
+        //those should'nt be here`
+
+        //// FIXME: 10/30/2016
```
Comment (teamagam3) [inline data/src/main/java/com/teamagam/gimelgimel/data/message/adapters/MessageDataMapper.java:192]: Fix it for the CR?

File: data/src/main/java/com/teamagam/gimelgimel/data/message/adapters/MessageDataMapper.java [hunk 11]
```diff
@@ -1,8 +1,10 @@
 package com.teamagam.gimelgimel.data.message.adapters;
```
Comment (teamagam3) [inline data/src/main/java/com/teamagam/gimelgimel/data/message/adapters/MessageDataMapper.java:1]: The subclasses in this file have more code lines then the main class, seperate them to different files.

File: data/src/main/java/com/teamagam/gimelgimel/data/map/adapter/GeoEntityDataMapper.java [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+import com.teamagam.gimelgimel.domain.map.entities.mapEntities.PointEntity;
+import com.teamagam.gimelgimel.domain.map.entities.symbols.PointSymbol;
+import com.teamagam.gimelgimel.domain.map.repository.GeoEntitiesRepository;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+/**
+ * Created on 11/1/2016.
+ * TODO: complete text
```
Comment (teamagam3) [inline data/src/main/java/com/teamagam/gimelgimel/data/map/adapter/GeoEntityDataMapper.java:16]: Complete the comment or remove it.

File: data/src/main/java/com/teamagam/gimelgimel/data/map/adapter/GeoEntityDataMapper.java [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    private GeoEntity createGeoEntity(String id, GeoContentData geoContentData) {
+        //todo: replace in the future with several types of GeoEntities
+        return new PointEntity(id, null,
+                (PointGeometry) mGeometryMapper.transform(geoContentData.getGeometry()),
+                new PointSymbol(geoContentData.getType(), geoContentData.getText()));
+    }
+
+    private class GeoContentToDataTransformer implements IGeoEntityVisitor {
```
Comment (teamagam3) [inline data/src/main/java/com/teamagam/gimelgimel/data/map/adapter/GeoEntityDataMapper.java:53]: The visitor could be the main class of this file.

File: data/src/main/java/com/teamagam/gimelgimel/data/location/adpater/LocationSampleDataAdapter.java [hunk 14]
```diff
@@ -0,0 +1,47 @@
+package com.teamagam.gimelgimel.data.location.adpater;
```
Comment (teamagam3) [inline data/src/main/java/com/teamagam/gimelgimel/data/location/adpater/LocationSampleDataAdapter.java:1]: This comment refers to all of the data adapters / mappers.

I think we should create a UniversalDataAdapter, it'll contain all of our adapters (List with interface or explicit members).
The adapter will get an object by method, and will decide which adapter should handle it,

**Review by teamagam3 — CHANGES_REQUESTED** [review]
